README.md
<a href="https://github.com/OWASP/owasp-masvs/releases/download/v1.2/OWASP_MASVS-v1.2-en.pdf"><img width="25%" align="right" style="float: right;" src="Document/images/masvs-mini-cover.png"></a>

# OWASP Mobile Application Security Verification Standard [![Twitter Follow](https://img.shields.io/twitter/follow/OWASP_MSTG.svg?style=social&label=Follow)](https://twitter.com/OWASP_MSTG)

[![Creative Commons License](https://licensebuttons.net/l/by-sa/4.0/88x31.png)](https://creativecommons.org/licenses/by-sa/4.0/ "CC BY-SA 4.0")
[![OWASP Flagship](https://img.shields.io/badge/owasp-flagship%20project-48A646.svg)](https://owasp.org/projects/)
[![Document Build](https://github.com/OWASP/owasp-masvs/workflows/Document%20Build/badge.svg)](https://github.com/OWASP/owasp-masvs/actions?query=workflow%3A%22CI+Build%22)
[![Check Markdown Markup](https://github.com/OWASP/owasp-masvs/workflows/Check%20Markdown%20Markup/badge.svg)](https://github.com/OWASP/owasp-masvs/actions?query=workflow%3A%22Check+Markdown+markup%22)
[![Check Markdown Links](https://github.com/OWASP/owasp-masvs/workflows/Check%20Markdown%20Links/badge.svg)](https://github.com/OWASP/owasp-masvs/actions?query=workflow%3A%22Check+Markdown+Links%22)

This is the official Github Repository of the OWASP Mobile Application Security Verification Standard (MASVS). The MASVS establishes baseline security requirements for mobile apps that are useful in many scenarios, including:

- In the SDLC - to establish security requirements to be followed by solution architects and developers;
- In mobile app penetration tests - to ensure completeness and consistency in mobile app penetration tests;
- In procurement - as a measuring stick for mobile app security, e.g. in form of questionnaire for vendors;
- Et cetera.

The MASVS is a sister project of the [OWASP Mobile Security Testing Guide](https://github.com/OWASP/owasp-mstg "OWASP Mobile Security Testing Guide").

# Getting the MASVS

PDF/Mobi/Epub/Docx downloads are available on the [Releases page](https://github.com/OWASP/owasp-masvs/releases "Releases"). The documents were created by using [pandocker](https://github.com/dalibo/pandocker/ "pandocker").

The current release of the [MASVS is version 1.2](https://github.com/OWASP/owasp-masvs/releases/tag/v1.2 "MASVS version 1.2"). Want to have the latest `snapshot` version? Check [the latest Github build action](https://github.com/OWASP/owasp-masvs/actions?query=workflow%3A%22Document+Build%22). The MASVS is also available in different languages:

- [Brazilian Portuguese (in progress)](https://github.com/OWASP/owasp-masvs/tree/master/Document-ptbr "Brazilian Portuguese")
- [Chinese (Simplified) - ZHCN](https://github.com/OWASP/owasp-masvs/tree/master/Document-zhcn "Simplified Chinese (ZHCN)")
- [Chinese (Traditional) - ZHTW](https://github.com/OWASP/owasp-masvs/tree/master/Document-zhtw "Traditional Chinese (ZHTW)")
- [English](https://github.com/OWASP/owasp-masvs/tree/master/Document "English")
- [French](https://github.com/OWASP/owasp-masvs/tree/master/Document-fr "French")
- [German](https://github.com/OWASP/owasp-masvs/tree/master/Document-de "German")
- [Hindi (in progress)](https://github.com/OWASP/owasp-masvs/tree/master/Document-hi "Hindi")
- [Japanese](https://github.com/OWASP/owasp-masvs/tree/master/Document-ja "Japanese")
- [Korean](https://github.com/OWASP/owasp-masvs/tree/master/Document-ko "Korean")
- [Persian](https://github.com/OWASP/owasp-masvs/tree/master/Document-fa "Persian")
- [Russian](https://github.com/OWASP/owasp-masvs/tree/master/Document-ru "Russian")
- [Spanish](https://github.com/OWASP/owasp-masvs/tree/master/Document-es "Spanish")  

## Gitbook

Read the English version on [Gitbook](https://mobile-security.gitbook.io/masvs/ "GitBook Mobile AppSec Verification Standard"). The book is automatically synchronized with the main repo.

## Create new PDF, Epub, Mobi, and Word document

You can find the documents in the release page. If you want to generate the documents yourself, execute the following steps. Clone the repository and run the document generator. This produces PDF, Epub and Docx files in the root of the project.

```shell
$ git clone https://github.com/OWASP/owasp-masvs/
$ cd owasp-masvs/tools/docker
$ ./run_docker_masvs_generation_on_local.sh LATEST
```

## Exporting to JSON, XML and CSV

The repository contains a Python tool for converting the requirements into various formats. Clone the repo and run `export.py` from the tools folder.

```shell
export.py [-h] [--format {json,xml,csv}] [--lang {es/ru/en/fr/de/zhtw/ja}]
```

## Suggestions and Feedback

To report and error or suggest an improvement, please create an [issue](https://github.com/OWASP/owasp-masvs/issues "Github issues") or create a Pull Request.

# How to Contribute

The MASVS is an open source effort and we welcome contributions and feedback. If you want to contribute additional content, or improve existing content, we suggest that you first contact us on the OWASP MSTG Slack channel:

<https://owasp.slack.com/messages/project-mobile_omtg/details/>

You can sign up here:

[https://owasp-slack.herokuapp.com/](https://owasp-slack.herokuapp.com/)

Before you start contributing, please check our [contribution guide](https://github.com/OWASP/owasp-masvs/blob/master/CONTRIBUTING.md "Contribution Guide") which should get you started.

# Read Individual Sections of the MASVS Here

- [Foreword](Document/0x01-Foreword.md)
- [Frontispiece](Document/0x02-Frontispiece.md)
- [The Mobile Application Security Verification Standard](Document/0x03-Using_the_MASVS.md)
- [Assessment and Verification](Document/0x04-Assessment_and_Certification.md)
- [V1: Architecture, Design and Threat Modeling Requirements](Document/0x06-V1-Architecture_design_and_threat_modelling_requireme.md)
- [V2: Data Storage and Privacy Requirements](Document/0x07-V2-Data_Storage_and_Privacy_requirements.md)
- [V3: Cryptography Requirements](Document/0x08-V3-Cryptography_Verification_Requirements.md)
- [V4: Authentication and Session Management Requirements](Document/0x09-V4-Authentication_and_Session_Management_Requirements.md)
- [V5: Network Communication Requirements](Document/0x10-V5-Network_communication_requirements.md)
- [V6: Environmental Interaction Requirements](Document/0x11-V6-Interaction_with_the_environment.md)
- [V7: Code Quality and Build Setting Requirements](Document/0x12-V7-Code_quality_and_build_setting_requirements.md)
- [V8: Resiliency Against Reverse Engineering Requirements](Document/0x15-V8-Resiliency_Against_Reverse_Engineering_Requirements.md)
- [Appendix A: Glossary](Document/0x90-Appendix-A_Glossary.md)
- [Appendix B: References](Document/0x91-Appendix-B_References.md)
- [Release notes](CHANGELOG.md)


CHANGELOG.md
# Changelog

## V1.2 7 March 2020 - International Release

The following changes are part of release 1.2:

- Translation in simplified Chinese of the MASVS available.
- Change of title in MASVS book cover.
- Removed Mobile Top 10 and CWE from MSTG and merged to existing references in MASVS.

## V1.2-RC 5 October 2019 - Pre-release (English Only)

The following changes are part of release 1.2:

- Promoted to flagship status.
- Requirement changed: MSTG-STORAGE-1 "need to be used".
- Requirements MSTG-STORAGE-13, MSTG-STORAGE-14, and MSTG-STORAGE-15 are added with a focus on data protection.
- Requirement MSTG-AUTH-11 is updated to preserve contextual information.
- Requirement MSTG-CODE-4 is updated to cover more than just debugging.
- Requirement MSTG-PLATFORM-10 added to further secure usage of WebViews.
- Requirement MSTG-AUTH-12 added to remind developers of having authorizations implemented, especially in case of multi-user apps.
- Added a little more description on how the MASVS should be used given a risk assessment.
- Added a little more description on paid content.
- Requirement MSTG-ARCH-11 added to include a Responsible Disclosure policy for L2 applications.
- Requirement MSTG-ARCH-12 added to show application developers that relevant international privacy laws should be followed.
- Created a consistent style for all references in the English version.
- Requirement MSTG-PLATFORM-11 added to counter spying via third party keyboards.
- Requirement MSTG-MSTG-RESILIENCE-13 added to impede eavesdropping at an application.

## V1.1.4 4 July 2019 - Summit Edition

The following changes are part of release 1.1.4:

- Fix all markdown issues.
- Updates in the French and Spanish translations.
- Translated the changelog to Chinese (ZHTW) and Japanese.
- Automated verification of the the markdown syntax and reachability of the URLs.
- Added identification codes to the requirements, which will be included in the future version of the MSTG in order to find the recommendations and testcases easily.
- Reduced the repo size and added Generated to the .gitignore.
- Added a Code of Conduct & Contributing guidelines.
- Added a Pull-Request template.
- Updated the sync with the repo in use for hosting the Gitbook website.
- Updated the scripts to generate XML/JSON/CSV for all the translations.
- Translated the Foreword to Chinese (ZHTW).

## V1.1.3 9 January 2019 - Small Fixes

- Fix translation issue of requirement 7.1 in the Spanish version
- New setup of translators in acknowledgements

## V1.1.2 3 January 2019 - Sponsorship and Internationalization

The following changes are part of release 1.1.2:

- Added thank you note for buyers of the e-book.
- Added missing authentication link & updated broken authentication link in V4.
- Fixed swap of 4.7 and 4.8 in English.
- First international release!
  - Fixes in Spanish translation. Translation is now in sync with English (1.1.2).
  - Fixes in Russian translation. Translation is now in sync with English (1.1.2).
  - Added first release of Chinese (ZHTW) French, German, and Japanese!
- Simplified document for ease of translation.
- Added instructions for automated releases.

## V1.1.0 14 July 2018

The following changes are part of release 1.1:

- Requirement 2.6 "The clipboard is deactivated on text fields that may contain sensitive data." was removed.
- Requirement 2.2 "No sensitive data should be stored outside of the app container or system credential storage facilities." was added.
- Requirement 2.1 was reworded to "System credential storage facilities are used appropriately to store sensitive data, such as PII, user credentials or cryptographic keys.".

## V1.0 12 January 2018

The following changes are part of release 1.0:

- Delete 8.9 as the same as 8.12
- Made 4.6 more generic
- Minor fixes (typos etc.)


tools/genpdf.js
/**
 * This is a first attempt in making a PDF using markdownpdf .
 * Please note that it is done in my spare time as a quick prototype: tackling one issue after another
 * Feel free to rewrite it as the core of this project is the actual resulting PDF, not necessary this javascript.
 * Author: Jeroen Willemsen
 */
var markdownpdf = require("markdown-pdf"),
  fs = require("fs"),
  split = require("split"),
  through = require("through"),
  duplexer = require("duplexer");
var Remarkable = require("remarkable");
var replace = require("replace-in-file");
var toc = require("markdown-toc");
var fs = require("fs");

var lang = "";
var langdir = "";
var tag = "";
var help = false;
var releaseNotes = "";

if (process.argv.includes("-h")) {
  console.log("Helptext here");
  help = true;
}
if (process.argv.includes("-lang")) {
  lang = process.argv[process.argv.indexOf("-lang") + 1];
} else if (process.argv.includes("-l")) {
  lang = process.argv[process.argv.indexOf("-l") + 1];
}
if (process.argv.includes("-tag")) {
  tag = process.argv[process.argv.indexOf("-tag") + 1];
}
if (tag != "") {
  tag = tag + " ";
}
tag = tag + "Date: " + setDate();

if (process.argv.includes("-relnotes")) {
  releaseNotes = process.argv[process.argv.indexOf("-relnotes") + 1];
} else {
  releaseNotes = generateRelNotes();
}

if (!help) {
  if (lang == "" || lang == null) {
    lang = "EN";
    langdir = "../Document/";
  } else {
    langdir = "../Document-" + lang + "/";
  }
  console.log("printing for " + lang);
  runPDF();
}

function preProcessRunningJs() {
  var options = {
    files: "running.js",
    from: "[DATE]",
    to: "[" + tag + "]"
  };
  try {
    const changes = replace.sync(options);
    console.log("Modified files:", changes.join(", "));
  } catch (error) {
    console.error("Error occurred:", error);
  }
}

function postProcessRunningJS() {
  const options = {
    files: "running.js",
    from: "[" + tag + "]",
    to: "[DATE]"
  };
  try {
    const changes = replace.sync(options);
    console.log("Modified files:", changes.join(", "));
  } catch (error) {
    console.error("Error occurred:", error);
  }
}

function generateRelNotes(){

  return fs.readFileSync("../CHANGELOG.md", "utf8");
}

function preProcessMd() {
  // Split the input stream by lines
  var splitter = split();
  var tocRendered = getToc();
  var replacer = through(function(data) {
    this.queue(
      data
        .replace("[date]", tag)
        .replace("[RELEASE_NOTES]", releaseNotes)
        .replace("[TOCCCC]", tocRendered)
        .replace("Images/", langdir + "/Images/") + "\n"
    );
  });

  splitter.pipe(replacer);
  return duplexer(splitter, replacer);
}

function setDate() {
  var today = new Date();
  var dd = today.getDate();
  var mm = today.getMonth() + 1; //January is 0!
  var yyyy = today.getFullYear();

  if (dd < 10) {
    dd = "0" + dd;
  }

  if (mm < 10) {
    mm = "0" + mm;
  }

  return mm + "/" + dd + "/" + yyyy;
}

function getPreDocSet() {
  return [
    langdir + "0x00-Header.md", 
    langdir + "Foreword.md",
    langdir + "SUMMARY.md"
  ];
}
function getPostDocSet() {
  return [
    //"changelog_and_toc.md",
    langdir + "0x02-Frontispiece.md",
    langdir + "0x03-Using_the_MASVS.md",
    langdir + "0x04-Assessment_and_Certification.md",
    langdir + "0x06-V1-Architecture_design_and_threat_modelling_requireme.md",
    langdir + "0x07-V2-Data_Storage_and_Privacy_requirements.md",
    langdir + "0x08-V3-Cryptography_Verification_Requirements.md",
    langdir + "0x09-V4-Authentication_and_Session_Management_Requirements.md",
    langdir + "0x10-V5-Network_communication_requirements.md",
    langdir + "0x11-V6-Interaction_with_the_environment.md",
    langdir + "0x12-V7-Code_quality_and_build_setting_requirements.md",
    langdir + "0x15-V8-Resiliency_Against_Reverse_Engineering_Requirements.md",
    langdir + "0x90-Appendix-A_Glossary.md",
    langdir + "0x91-Appendix-B_References.md"
  ];
}
function getDocSet() {
  return getPreDocSet().concat(getPostDocSet());
}

function getToc() {
  var resultString = "";
  var docSet = getPostDocSet();
  docSet.forEach(function(filename) {
    resultString = resultString + fs.readFileSync(filename, "utf8");
  });
  var renderedToc = toc(resultString, { maxdepth: 3 }).content;
  return renderedToc;
}

function runPDF() {
  var mdDocs = getDocSet(),
    bookPath = "./../generated/MSTG-" + lang + ".pdf";

  preProcessRunningJs();
  markdownpdf({
    preProcessMd: preProcessMd,
    remarkable: {
      html: true,
      breaks: true
    },
    runningsPath: "running.js",
    cssPath: "pdf.css"
  })
    .concat.from(mdDocs)
    .to(bookPath, function() {
      console.log("Created", bookPath);
      postProcessRunningJS();
    });
}
